### ISSUE TITLE
[Build] compilation error: invalid instruction mnemonic 'vcvtneeph2ps'

### ISSUE THREAD

**Issue by saiden89** [microsoft/onnxruntime#22519]
### Describe the issue

I am attempting to compile ONNX Runtime on the [LUMI](https://docs.lumi-supercomputer.eu/) supercomputer, a Cray system. 

The configuration step is completed without any issues. However, during the compile phase, I encountered problems when using the default `CC` and `cc` [Cray compiler wrappers](https://docs.lumi-supercomputer.eu/development/compiling/prgenv/#compiler-wrappers), which apply Cray-specific optimizations. To bypass this, I manually specified the AMD compilers (`amdclang` and `amdclang++`) instead of the wrappers.

### System Details:
- **GPU**: AMD MI250X (`gfx90a`)
- **CPU**: AMD EPYC 7A53 "Trento"

Now, I’m encountering a compile-time error possibly related to the AVX512 instruction set: `error: invalid instruction mnemonic 'vcvtneeph2ps'`, but I’m not familiar enough with all this to diagnose the issue. I would appreciate any guidance on how to address this.

### Urgency

Not urgent, but would be nice to have since I have a big inference job on a project.

### Target platform

AMD MI250X

### Build script

The build script relies on some specific modules being loaded to target the correct architecture, as well as loading the correct [programming environment](https://docs.lumi-supercomputer.eu/development/compiling/prgenv/). Full reproducibility might be limited because of the exotic nature of the system, but I am more than happy to try myself any suggestions.

```bash
module purge

module load PrgEnv-amd
module load rocm/6.0.3
module load craype-accel-amd-gfx90a craype-x86-trento

cd /tmp || exit
git clone --single-branch --branch main --recursive https://github.com/Microsoft/onnxruntime onnxruntime
cd onnxruntime || exit

mamba install rust -y
pip install cmake

./build.sh --config Release \
    --build_wheel \
    --update \
    --build \
    --parallel \
    --use_rocm \
    --rocm_home "$ROCM_PATH" \
    --cmake_extra_defines CMAKE_HIP_ARCHITECTURES=gfx90a \
    --cmake_extra_defines CMAKE_C_COMPILER=amdclang \
    --cmake_extra_defines CMAKE_CXX_COMPILER=amdclang++

pip install build/Linux/Release/dist/*
```

### Error / output

[log.txt](https://github.com/user-attachments/files/17458861/log.txt)

### GCC / Compiler Version

AMD clang version 17.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-6.0.3 24012 af27734ed982b52a9f1be0f035ac91726fc697e4)

**Comment by edgchen1** [comment]
Here's the error from the log for convenience:
```
Building ASM object CMakeFiles/onnxruntime_mlas.dir/tmp/onnxruntime/onnxruntime/core/mlas/lib/x86_64/cvtfp16Avx.S.o
/tmp/onnxruntime/onnxruntime/core/mlas/lib/x86_64/cvtfp16Avx.S:60:9: error: invalid instruction mnemonic 'vcvtneeph2ps'
        vcvtneeph2ps ymm0, ymmword PTR [rdi]
        ^~~~~~~~~~~~
```

I think this code was added in this PR:
https://github.com/microsoft/onnxruntime/pull/21183

@eralmual do you have any pointers on how to fix this?

**Comment by eralmual** [comment]
Hi! Thank you for reaching out!

Seems like the _vcvtneeph2ps_ instruction not recognized by the compiler, I did a quick search and [the instruction is supported on Clang since v16.0](https://releases.llvm.org/16.0.0/tools/clang/docs/ReleaseNotes.html) as part of the AVX-NE-CONVERT ISA, seems like you are using v17.0 so it should work fine.

If it's not working for Clang in general I can do a quick patch to prevent the compiler error while we find a solution, just let me know. 
In the meanwhile i think you should be able to safely delete the if and everything inside at line https://github.com/microsoft/onnxruntime/blob/c7138a2630b01e30340a52959c232305394fd86f/cmake/onnxruntime_mlas.cmake#L574 and that should fix the compiler issue. 

Let me know if it works!

**Comment by snnn** [comment]
It is more about if your Assembler(like [gas](https://en.wikipedia.org/wiki/GNU_Assembler)) can recognize this instruction. We should write a test program to check it: https://cmake.org/cmake/help/latest/module/CheckSourceCompiles.html, instead of detecting compiler name/version. 

Contributions are welcomed

**Comment by saiden89** [comment]
Thank you @eralmual  for the suggestion, your proposed solution solves the problem. However, as the compilation continues I am greeted by a lot more errors.
```shell
/pfs/lustrep2/projappl/project_465000941/compartments/onnxruntime/build/Linux/Release/amdgpu/onnxruntime/core/providers/rocm/tensor/cast_op.cc:295:1: error: explicit instantiation of 'ComputeInternal' that occurs after an explicit specialization has no effect [-Werror,-Winstantiation-after-specialization]
SPECIALIZE_IMPL(MLFloat16)

/pfs/lustrep2/projappl/project_465000941/compartments/onnxruntime/onnxruntime/core/providers/rocm/nn/conv_impl.cu:24:21: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  fast_divmod fdm_c(bias_size);
              ~~~~~ ^~~~~~~~~
```
Any further insights are deeply appreciated, thanks!

**Comment by snnn** [comment]
Please add "--compile_no_warning_as_error" to your build command.

**Comment by snnn** [comment]
We don't use clang to build our CUDA code. Therefore we didn't see such warnings.  You can help us fix them or suppress them if you'd like.  Contributions are welcome. Thanks.

**Comment by cocodee** [comment]
remove the if(CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL 13.1 AND NOT(APPLE)) 
and I see below error
2407.1 [ 67%] Linking CXX executable onnxruntime_mlas_q4dq
2407.4 /usr/bin/ld: libonnxruntime_mlas.a(platform.cpp.o): in function `MLAS_PLATFORM::MLAS_PLATFORM()':
2407.4 platform.cpp:(.text._ZN13MLAS_PLATFORMC2Ev+0x4d2): undefined reference to `MlasCastF16ToF32KernelAvx'
2407.4 collect2: error: ld returned 1 exit status
2407.4 gmake[2]: *** [CMakeFiles/onnxruntime_mlas_q4dq.dir/build.make:164: onnxruntime_mlas_q4dq] Error 1
2407.4 gmake[1]: *** [CMakeFiles/Makefile2:1938: CMakeFiles/onnxruntime_mlas_q4dq.dir/all] Error 2
2407.4 gmake[1]: *** Waiting for unfinished jobs....

**Comment by Jiuzun** [comment]
@cocodee I also in this issue, have you solved it?

**Comment by JensHuthmann** [comment]
I get the same problem on `AMD EPYC 9554` cpu using gcc
`g++-13 (Ubuntu 13.1.0-8ubuntu1~22.04) 13.1.0`

cmake command
`CC=/usr/bin/gcc-13 CXX=/usr/bin/g++-13 cmake ../cmake`

**Comment by MDLZCOOL** [comment]
@cocodee I encountered the same issue, did you solve it?

**Comment by BenLeadbetter** [comment]
Same here. There is the `onnxruntime_FORCE_GENERIC_ALGORITHMS` cmake option which would likely solve the issue if we turn it on, but it says "Use only for testing and debugging generic algorithms", also it would be a shame to disable other optimisations just becuase of a few missing instructions.

Ideally we'd make these checks smarter, and able to detect when the instructions are actually available. Currently [the offending code](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/mlas/lib/platform.cpp#L503-L510) is behind a `defined(__GNUC__) && (__GNUC__ >= 13)` check on linux which doesn't seem to be quite correct.

**Comment by BenLeadbetter** [comment]
I guess this patch might suffice to unblock people? 

```patch
diff --git a/cmake/onnxruntime_mlas.cmake b/cmake/onnxruntime_mlas.cmake
index 15864a0198..862f1ce0ce 100644
--- a/cmake/onnxruntime_mlas.cmake
+++ b/cmake/onnxruntime_mlas.cmake
@@ -602,12 +602,6 @@ else()
           ${MLAS_SRC_DIR}/rotary_embedding_kernel_avx2.cpp
           ${MLAS_SRC_DIR}/rotary_embedding_kernel_avx2_fp32.cpp
         )
-        if(CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL 13.1 AND NOT(APPLE))
-          set(mlas_platform_srcs_avx2
-            ${mlas_platform_srcs_avx2}
-            ${MLAS_SRC_DIR}/x86_64/cvtfp16Avx.S
-          )
-        endif()
 
 message(STATUS "CMAKE_CXX_COMPILER_ID: ${CMAKE_CXX_COMPILER_ID}")
 message(STATUS "CMAKE_CXX_COMPILER_VERSION: ${CMAKE_CXX_COMPILER_VERSION}")
diff --git a/onnxruntime/core/mlas/lib/platform.cpp b/onnxruntime/core/mlas/lib/platform.cpp
index 582c1ab944..d09a65168c 100644
--- a/onnxruntime/core/mlas/lib/platform.cpp
+++ b/onnxruntime/core/mlas/lib/platform.cpp
@@ -500,7 +500,7 @@ Return Value:
                 }
 
 #ifndef __APPLE__
-#if (defined(_MSC_VER) && (_MSC_VER >= 1933)) || (defined(__GNUC__) && (__GNUC__ >= 13))
+#if (defined(_MSC_VER) && (_MSC_VER >= 1933))
                 //
                 // Check if the processor supports AVX NE CONVERT.
                 //
```

**Comment by snnn** [comment]
> Currently [the offending code](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/mlas/lib/platform.cpp#L503-L510) is behind a defined(__GNUC__) && (__GNUC__ >= 13) check on linux which doesn't seem to be quite correct.

I agree with you.

**Comment by horenmar** [comment]
I ran into this issue when compiling ORT 1.22.1 for plain Ubuntu Jammy (22.04). For an interesting datapoint, we use the same compiler on all platforms - GCC 13.4.0, but the build succeeds on newer Ubuntus, while failing only on Jammy..

**Comment by snnn** [comment]
It is not about GCC's version. It is about [GAS](https://en.wikipedia.org/wiki/GNU_Assembler)'s version.

**Comment by PriscillaJCorn** [comment]
> It is more about if your Assembler(like [gas](https://en.wikipedia.org/wiki/GNU_Assembler)) can recognize this instruction. We should write a test program to check it: https://cmake.org/cmake/help/latest/module/CheckSourceCompiles.html, instead of detecting compiler name/version.
> 
> Contributions are welcomed

@snnn I agree with you that we should check whether the compiler can recognize this instruction, instead of detecting compiler name/version. So, why not do some checks before we start, such as: 
`        include(CheckCSourceCompiles)

        set(CMAKE_REQUIRED_FLAGS "-mavx512fp16")
        check_c_source_compiles("
        int main() {
            __asm__ volatile(\"vcvtneeph2ps %ymm0, %ymm1\");
            return 0;
        }
        " COMPILER_SUPPORTS_AVX512FP16)


        if(COMPILER_SUPPORTS_AVX512FP16 AND NOT APPLE)
          set(mlas_platform_srcs_avx2
            ${mlas_platform_srcs_avx2}
            ${MLAS_SRC_DIR}/x86_64/cvtfp16Avx.S
          )
        endif()`   
This method helped me fix the compiler issue. Can I submit a PR to make some contributions？

**Comment by snnn** [comment]
Sure. Welcome to contribute!  (I no longer work on this project. Someone else will review it)

**Comment by PriscillaJCorn** [comment]
@snnn Thank you for your reply. I have submitted a PR (#28236) to address this issue. Could someone  help me with a review? @edgchen1 @horenmar
